frontend/src/features/map_planning/layers/shade/components/Shading.tsx

```diff
@@ -0,0 +1,231 @@
+  );
+  const setSingleNodeInTransformer = useMapStore((store) => store.setSingleNodeInTransformer);
+  const addNodeToTransformer = useMapStore((store) => store.addNodeToTransformer);
+  const removeNodeFromTransformer = useMapStore((store) => store.removeNodeFromTransformer);
+  const executeAction = useMapStore((store) => store.executeAction);
+
+  const geometry = shading.geometry as PolygonGeometry;
+
+  const isShadingEdited =
+    shadingManipulationState !== 'inactive' &&
+    selectedShadings?.length === 1 &&
+    selectedShadings[0].id == shading.id;
+
+  const removeShadingFromSelection = (e: KonvaEventObject<MouseEvent>) => {
+    const selectedShadings = (foundShadings: ShadingDto[], konvaNode: Node) => {
+      const shadingNode = konvaNode.getAttr('shading');
+      return shadingNode ? [...foundShadings, shadingNode] : [foundShadings];
+    };
+
+    const getUpdatedShadingSelection = () => {
+      const transformer = useMapStore.getState().transformer.current;
+      return transformer?.nodes().reduce(selectedShadings, []) ?? [];
+    };
+
+    removeNodeFromTransformer(e.currentTarget);
+    selectShadings(getUpdatedShadingSelection());
+  };
+
+  const addShadingToSelection = (e: KonvaEventObject<MouseEvent>) => {
+    addNodeToTransformer(e.currentTarget);
+
+    const currentPlantingSelection =
+      useMapStore.getState().untrackedState.layers.shade.selectedShadings ?? [];
+    selectShadings([...currentPlantingSelection, shading]);
+  };
+
+  const handleMultiSelect = (e: KonvaEventObject<MouseEvent>, shading: ShadingDto) => {
+    isShadingElementSelected(shading) ? removeShadingFromSelection(e) : addShadingToSelection(e);
+  };
+
+  const handleSingleSelect = (e: KonvaEventObject<MouseEvent>, shading: ShadingDto) => {
+    setSingleNodeInTransformer(e.currentTarget);
+    selectShadings([shading]);
+  };
+
+  const handleClickOnShading = (e: KonvaEventObject<MouseEvent>) => {
+    if (isShadingPlacementModeActive() || shadingManipulationState !== 'inactive') return;
+
+    isUsingModifierKey(e) ? handleMultiSelect(e, shading) : handleSingleSelect(e, shading);
+  };
+
+  useEffect(() => {
+    stageListenerRegister.registerStageClickListener(`Shading-${shading.id}`, (e) => {
+      if (!isShadingEdited || shadingManipulationState !== 'add') return;
+
+      const newPoint = {
+        x: e.currentTarget.getRelativePointerPosition().x,
+        y: e.currentTarget.getRelativePointerPosition().y,
+        srid: DEFAULT_SRID,
+      };
+
+      const newGeometry = insertPointIntoLineSegmentWithLeastDistance(
+        geometry,
+        newPoint,
+        editorLongestSide / 100,
+      );
+
+      executeAction(
+        new UpdateShadingAction({
+          id: shading.id,
+          shade: shading.shade,
+          geometry: newGeometry as object,
+        }),
+      );
+    });
+  }, [shadingManipulationState, isShadingEdited, shading, geometry, editorLongestSide]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
This prevents infinite loops caused by some functions being updated after a rerender of the component.
Again, I have to admit this is mostly because of poor design decisions and little knowledge about react on my part.